Add initial draft of the SE-Connect contribution document. (#117)
ShipEngine/connect: samples/contributing.md

> # Contributing (connect apps)

Author: Updated all instances to use `Connect`.

> When contributing a connect app to the SE-C monorepo please follow the guidelines outlined in this document. Adherence to this set of standards will help to ensure that both the contributor as well as the reviewer have similar expectations related to the code contained herein.

Reviewer: ```suggestion
When developing a ShipEngine Connect app, please follow the guidelines outlined in this document. Adherence to this set of standards will help to ensure that both the contributor as well as the reviewer have similar expectations related to the code contained herein.
```

   ```bash
   # Scaffold the app.
   connect init
   ...[answer the connect generator questions]...
   # Initialize a git repo.
   git init
   # Stage the scaffolded file to be tracked by git.
   git add .
   # Commit the repo in it's new, prestine state.
   git commit -m "Initial commit, scaffolding is now in place... no other changes."
   ```

4. Open a PR at this point so we can get the baseline code in place to iterate from thereafter.

5. The repo is now ready for you to begin implementing your connect application.

## Best practices

### General guidance

1. The documentation site can be found here, http://connect.shipengine.com/. This is the best single resource to answer your questions related to building a connect app. Please familiarize yourself with these docs.
   - It will be helpful to review the [Sample Apps](https://github.com/ShipEngine/connect-samples) as well to get a grasp on how implementation can be handled in a few example scenarios.
2. Do not use non-public shared packages, as they will fail to build and deploy properly inside our production environments.

>    - If a package is **not** publicly hosted and available via `npm`, it should be not referenced as a dependency within your app.

Reviewer 2: related to this is that they can't reference anything outside of that main directory

Author: Added language as suggested.

> 3. Naming is important when defining the `npm` package... as this is used to map the application to the infrastructure that host's it.

Reviewer 2: ```suggestion
3. Naming is important when defining the `npm` package... as this is used to map the application to the infrastructure that hosts it.
```

3. Clear communication is essential and expected throughout the process.
4. Feedback on how the offering can be improved to facilitate better development must also be a primary consideration.

### When the docs are not enough

1. When questions arise which are not addressed within the documentation, they should be directed to the Engineering Coordinators to effectively answer the questions and work with the team to update the Connect SDK and/or documentation to clarify the situation. 
2. Likewise, when there are features or elements that are supported within a Carrier or Order Source which do not correlate with any available methods in ShipEngine Connect, we would like to be made aware of them so that we can investigate and potentially fill the noted gaps in supported functionality.

### Code quality

1. **Security is paramount**. As such there is a lot of time and thought that should be taken to ensure both the code as well as the process is secure. Listed below are some common considerations that should be taken to help ensure security gaps are considered and mitigated during the development phase.

   - Never store credentials inside the git repository.

>    - Frequently `client_id` and `client_secret`'s are needed to allow an integration to authenticate with 3rd party api's. The secret should always remain secret.

Reviewer 2: ```suggestion
   - Frequently `client_id` and `client_secret`s are needed to allow an integration to authenticate with 3rd party apis. The secret should always remain secret.
```

       - Note that when using HTTPS, even the query string will be protected, but the path is likely to still end up inside the logs.

2. Audit your commit history to ensure you have not committed sensitive data into the version control history.

   - `git grep SENSITIVE_DATA_STRING_HERE $(git rev-list --all)`
     - This will search through your git history attempting to find instances of the sensitive data string.
       - Should you get matching results, you should be sure to modify the git history to remove the values.

3. Secure access to the git repo.

   - Require 2FA.
   - Never use credentials that are shared across multiple users.
   - Contributors should only have access to the data they need to do their work.
   - GitHub accounts are commonly personal accounts... be sure to remove access for users when they leave the company.

4. Do not make assumptions in code that could be hard to change later.

### Code style

We suggest that you use the latest ES6 conventions.

> 1. Name things logically... [See reference](https://www.nikola-breznjak.com/blog/books/programming/code-complete-2-steve-mcconnell-power-variable-names/).

Reviewer 2: My code now:
```
let logically = 'foo'
...
logically = true
...
logically = 6
```

2. Do not use `eval`. [See reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#never_use_eval!).

3. Do not use `with`. [See reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/with#ambiguity_contra).

4. Do not use primitive type wrapper constructors: `new Number`, `new String`, `new Boolean`, `new Array`, `new Object`

   - Primitive type wrappers are slower than their primitive counterparts.

   - They also cause subtle bugs with `instanceof`.

>    - Object should be created using using the object literal syntax.

Reviewer 2: Never say never. I prefer `Boolean(foo)` over `!!isFoo`.

Author: See items 14, Do not use forced boolean casting => `if(!!value)`.
But to your point, readability is also a consideration that must be taken... I have updated the language around this.

Reviewer 2: Cool, missed that

       ```js
       // Bad.
       const a = {
         k1 : v1,
         k2: v2
       };
       const b = { k1: v1, k2 : v2, };
       const c = {k1:v1};
       
       // Good.
       const a = {
         k1: v1,
         k2: v2,
       };
       const b = { k1: v1, k2: v2 };
       const c = { k1: v2 };
       ```

   - Arrays should be created using the array literal syntax.

     ```js
     // Bad.
     const a = new Array();
     
     // Good.
     const a = [];
     ```

6. Favor `const` and `let` over `var`.

   - `var` is a legacy keyword and there are no practical use cases for `var` in modern ES6.

     - Never use `var`.
   - `const` is the default. This prevents the variable from being re-assigned, however a `const` object can still be mutated; if you need true immutability, consider using [`immutable-js`](http://facebook.github.io/immutable-js/).
   - `let` only when the value of the variable must be re-assigned. This is allowed, but it is commonly a sign that you should refactor the code to prevent the need to use `let`.

7. When applicable, use `for..of` and `for..in` instead of `for..i`.

>    - I most cases you should favor `forEach` and `map`.

Reviewer 2: ```suggestion
   - In most cases you should favor `forEach` and `map`.
```

    }
    ```

12. Named params are preferred, however if varaidic expansion is required... always favor the ES6 `...rest` syntax over inspecting the `arguments` object.

13. Options params should be passed in via an object... instead of as a named param.

    - This allows the options to be passed with differing data without changing how the function is called.

    - Additionally it helps to prevent truthiness bugs when booleans are passed. See [boolean trap anti-pattern](https://ariya.io/2011/08/hall-of-api-shame-boolean-trap).

      ```js
      // Bad.
      function apiCall(a, b, isMock = false ) { }
      
      // Good.
      function apiCall(a, b, { isMock = false } = {}) {}
      ```

14. Dot notation usage should be consistent.

    ```js
    // Bad.
    obj.x = 123;
    obj['y'] = 321;
    
    // Ok, if needed.
    obj['x'] = 123;
    obj['y-z'] = 321;
    
    // Best.
    obj.x = 123;
    obj.y = 321;
    ```

15. Always use strict comparison operators.

    - Do not use `!=` and `==`.
    - Use these instead `!==` and `===`.
    - Prefer explicit strict value checks over truthiness checks.
    - Do not use forced boolean casting => `if(!!value)`.

> 16. Method chains must start with a leading dot.

Reviewer 2: Prefer `async/await` over `Promise`?

Author: I have added an item to state that the async / await pattern is preferred over the promise pattern.

    ```js
    // Bad.
    myPromise.then(() => {}).
      then(() => {}).
      catch((ex) => {})
    
    // Good.
    myPromise
      .then(() => {})
      .catch((ex) => {})
    ```

17. Prefer the **nullish coalescing operator (`??`)** over explicit `null` / `undefined` / truthy checks.

    ```js
    const y = null;
    
    // Bad
    const x = (y === undefined || y === null || !y) ? "default value" : y;
    
    // Good.
    const x = y ?? "default value"
    ```

18. Do not nest ternary operators `(?:)`. This is a readability issue.

19. Favor `import { myFunc } from './path/to/module';` instead of `const myFunc = require('./path/to/module');`

20. Do not use the `delete` keyword unless you have to... Instead prefer setting the value of the key to `null` instead. This is a performance consideration.

### Commits

> 1. Be sure your [git config](https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration) is properly setup... This will help to ensure that changes can be accuratley be traced back to the author who made the change.

Reviewer 2: ```suggestion
1. Be sure your [git config](https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration) is properly setup... This will help to ensure that changes can be accurately be traced back to the author who made the change.
```

2. Work inside branches... do not commit code directly to `master`.
3. Each and every commit message should contain a meaningful description of the changes made within that commit.
   - Bad: `Refactored entire app as requested.`
   - Good: `Updated yarn version from '1.17.0' to version '1.22.10', so we no longer get errors when node has not been installed as root.`
   - Note that a good commit message should describe both what was changed, as well as why the change was needed.
4. Break up the commits into logical chunks.  Each commit should be self contained, such that a single commit can be reverted and it will revert all that is needed to get back to a functional state.
   - For example, if a dependency is updated which required code changes to take into use... You should include the update as well as the changed needed to make it function.

>      -  These changes can be included within a single commit if the PR contains more changes than just this individual item.

Reviewer 2: ```suggestion
     - These changes can be included within a single commit if the PR contains more changes than just this individual item.
```

         - Application ID
2. Partners are responsible for peer review iterations of their own modules prior to requesting a review from Auctane.
   - Make those efforts transparent via Github.
   - This keeps our engineering out of the minutiae of code quality, and other obvious technical issues.
3. When creating a Pull Request (PR), you should be very clear and concise in both the `Title` as well as the `Summary / Description`.
   - A good title will breifly describe what the PR is intended to do or change.
   - A good description will provide a concise breakdown of what was changed and why... In addition any additional reference information and context should be linked or explicitly provided inside the PR summary.
4. A PR should not be huge. Much like well designed code, a well made PR should adhere to a single responsibility principal.  Multiple PRs with a clear seperation of responsibilities between each are far better than a huge PR with multiple changes with varying purposes throughout.
   - Prefer multiple PRs with a single responsibility instead of a single PR with the whole app's implementation for review.
5. Once review comments have been made on a PR:

>    - For explicitly requested changes... once you have made the requested change, you should mark the converstation as resolved in the PR.

Reviewer 2: Question of how much we can trust contractors... I've been glad some contractors don't mark them as resolved because I have to go through and make sure they actually acted on the comment.

Author: I have updated the language to ensure that the change has been made via the tooling's suggested changes feature... without need for manual verification.

>      - IE, Rename method `getForDisplay` to `getRecipientNameForDisplay`.

Reviewer 2: ```suggestion
     - I.e., rename method `getForDisplay` to `getRecipientNameForDisplay`.
```

   - For more general requested changes, you should make the changes and await or request a re-review. Once the reviewer has signed off on the changes, the reviewer will mark the conversation as resolved.

>      - IE, Refactor the api call to use the XML endpoint instead of the JSON endpoint when calling the API... there are know issues with the JSON endpoint.

Reviewer 2: ```suggestion
     - I.e., refactor the api call to use the XML endpoint instead of the JSON endpoint when calling the API... there are know issues with the JSON endpoint.
```

# Contributing (Connect apps)

> When contributing a Connect app to the SE-C monorepo please follow the guidelines outlined in this document. Adherence to this set of standards will help to ensure that both the contributor as well as the reviewer have similar expectations related to the code contained herein.

Reviewer: ```suggestion
When contributing a Connect app to the ShipEngine Connect platform, please follow the guidelines outlined in this document. Adherence to this set of standards will help to ensure that both the contributor as well as the reviewer have similar expectations related to the code contained herein.
```

## Getting started

1. To get started with Connect, follow the [online documentation to create a new app](https://connect.shipengine.com/docs/cli#initialize-a-new-app). 

> 2. **TODO** - Add specific details related to the `connect init` generator questions.

Author: Done.

>    - Since the Connect generator is about to be updated to work differently that it currently does, I am leaving this as a **TODO** item.

Author: ```suggestion
   - Since the Connect generator is about to be updated to work differently than it currently does, I am leaving this as a **TODO** item.
```

3. As a best practice to help ensure that visibility is clear for any changes made after the template has been scaffolded on your local machine, we ask that you create an initial commit before adding any custom changes.

   ```bash
   # Scaffold the app.
   connect init
   ...[answer the Connect generator questions]...
   # Initialize a git repo.
   git init
   # Stage the scaffolded file to be tracked by git.
   git add .

>    # Commit the repo in it's new, prestine state.

Reviewer: ```suggestion
   # Commit the repo in its new, pristine state.
```

   git commit -m "Initial commit, scaffolding is now in place... no other changes."
   ```

> 4. Open a PR at this point so we can get the baseline code in place to iterate from thereafter.

Reviewer: ```suggestion
4. Open a pull request (PR) in GitHub at this point so we can get the baseline code in place to iterate from thereafter.
```

   - If a package is **not** publicly hosted and available via `npm`, it should be not referenced as a dependency within your app.
   - This is due to the fact that your app will be unable to reference anything that resides outside of that root application directory (where the package.json is found).
3. Naming is important when defining the `npm` package... as this is used to map the application to the infrastructure that host's it.
   - Be sure to properly define the package name as well as the company name.
     - Example, the `Acme` company is developing the `Ship-Right` carrier application.
       - The company name / scope should be: `@acme`
       - The application name should be: `ship-right`
4. Keep your `connect-sdk` dependency up to date. We publish updates frequently and will have better development experience and will have an easier review and problem solving experience if you are using the latest available sdk version.

### Expectations for contractors

1. Deliver a functional application that fulfills the outlined requirements within an established timeframe. 

> 2. As development progresses, status updates and communication of any barriers should happen without delay so they can be address in a timely manner.

Reviewer 3: minor typo - "so they can be addressed in a timely manner."

Author: Fixed.

> 3. Clear communication is essential to deliverying a high quality integration and should be done throughout the process.

Reviewer 3: another minor typo  - "deliverying"

Author: Fixed.

       - Note that when using HTTPS, even the query string will be protected, but the path is likely to still end up inside the logs.

2. Audit your commit history to ensure you have not committed sensitive data into the version control history.

   - `git grep SENSITIVE_DATA_STRING_HERE $(git rev-list --all)`
     - This will search through your git history attempting to find instances of the sensitive data string.
       - Should you get matching results, you should be sure to modify the git history to remove the values.

3. Secure access to the git repo.

   - Require 2FA.
   - Never use credentials that are shared across multiple users.
   - Contributors should only have access to the data they need to do their work.
   - GitHub accounts are commonly personal accounts... be sure to remove access for users when they leave the company.

4. Do not make assumptions in code that could be hard to change later.

### Code style

We suggest that you use the latest ES6 conventions.

> 1. Name things logically... [See reference](https://www.nikola-breznjak.com/blog/books/programming/code-complete-2-steve-mcconnell-power-variable-names/).

Reviewer: Consider moving the details of this section into a separate file and link to it so that the section below this have better visibility

Author: Extracted to a separate doc and referenced from this one.

     - Or alternatively in seperate commits but both inside a PR  related to this specific change.
     - Basically, commit related work together.
5. Do not re-write the history of published branches, like `release ` / `master`.
   - Git rebasing is fine for your topic branch, but not for published branches, like `release` / ` master` branches.
6. Keep your working branch up-to-date with `master`.  When working inside a long lived topic branch, it is easy to find that your branch has gone out-of-date with the `master` or branch you originally based from.
   - Writing and testing new code against a branch that is out of date is not useful since you cannot be sure that it will function properly once merged.
   - You should frequently rebase against or merge the upstream branch into your local branch often.

## Pull Request Process

1. `README.md` - Always include a readme file that clearly describes the requirements of the integration, as well as how to get the app up and running after cloning it from the git repository.
   - README sections should include but are not limited to:
      - Link to the website of the company being integrated with.

>       - Links to the relevant documentation.

Reviewer: We should encourage developers to include documentation files as well in a `/docs` directory, when possible

Author: Added additional entry on the next line noting your suggest from above.
